format/bplist/bplist.go

```diff
@@ -0,0 +1,245 @@
+package bplist
+
+import (
+	"math"
+	"time"
+
+	"github.com/wader/fq/format"
+	"github.com/wader/fq/pkg/decode"
+	"github.com/wader/fq/pkg/interp"
+	"github.com/wader/fq/pkg/scalar"
+)
+
+func init() {
+	interp.RegisterFormat(decode.Format{
+		Name:        format.BPLIST,
+		ProbeOrder:  format.ProbeOrderBinFuzzy, // after most others (overlap some with webp)
+		Description: "Apple Binary Property List",
+		Groups:      []string{format.PROBE},
+		DecodeFn:    bplistDecode,
+	})
+}
+
+const (
+	elementTypeNullOrBoolOrFill = 0x00
+	elementTypeInt              = 0x01
+	elementTypeReal             = 0x02
+	elementTypeDate             = 0x03
+	elementTypeData             = 0x04
+	elementTypeASCIIString      = 0x05
+	elementTypeUnicodeString    = 0x06
+	elementTypeUID              = 0x08
+	elementTypeArray            = 0x0A
+	elementTypeSet              = 0x0C
+	elementTypeDict             = 0x0D
+)
+
+const (
+	null      = 0x00
+	boolFalse = 0x08
+	boolTrue  = 0x09
+)
+
+var singleByteElementMap = scalar.UToScalar{
+	null:      {Sym: "null", Description: "Null Value"},
+	boolFalse: {Sym: "false", Description: "False Value"},
+	boolTrue:  {Sym: "true", Description: "True Value"},
+}
+
+var elementTypeMap = scalar.UToScalar{
+	elementTypeNullOrBoolOrFill: {Sym: "singleton", Description: "Singleton value (null/bool)"},
+	elementTypeInt:              {Sym: "int", Description: "Integer"},
+	elementTypeReal:             {Sym: "real", Description: "Floating Point Number"},
+	elementTypeDate:             {Sym: "date", Description: "Date, 8-byte float"},
+	elementTypeData:             {Sym: "data", Description: "Binary data"},
+	elementTypeASCIIString:      {Sym: "ascii_string", Description: "ASCII encoded string"},
+	elementTypeUnicodeString:    {Sym: "unicode_string", Description: "Unicode string"},
+	elementTypeUID:              {Sym: "uid", Description: "UID"},
+	elementTypeArray:            {Sym: "array", Description: "Array"},
+	elementTypeSet:              {Sym: "set", Description: "Set"},
+	elementTypeDict:             {Sym: "dict", Description: "Dictionary"},
+}
+
+// decodes the number of bits required to store the following object
+func decodeSize(d *decode.D) uint64 {
+	n := d.FieldU4("size_bits")
+	if n != 0x0F {
+		return uint64(n)
+	}
+
+	d.FieldU4("large_size_marker", d.AssertU(1)) // TODO: add assertion that this is 0001
+
+	// get the exponent value
+	n = d.FieldU4("exponent")
+
+	// calculate the number of bytes encoding the size
+	n = uint64(math.Pow(2, float64(n)))
+
+	// decode that many bytes as big endian
+	n = d.FieldUBigIntBE("size", int(n*8)).Uint64()
+	return n
+}
+
+func decodeItem(d *decode.D, p *plist) any {
+	d.FieldStruct("object", func(d *decode.D) {
```

Comment:
Is the object struct a bit redundant? looking a the decode tree maybe the item could decode directly into the key/value struct making the tree a bit less deep? but maybe the spec wants this structure?